File: java_framework/src/main/java/com/automation/base/BasePage.java
package com.automation.base;

import org.openqa.selenium.*;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.io.File;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

import org.apache.commons.io.FileUtils;

public class BasePage {

    protected WebDriver driver;
    protected WebDriverWait wait;
    protected Actions actions;

    public BasePage(WebDriver driver) {
        this(driver, 20);
    }

    public BasePage(WebDriver driver, int waitTime) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(waitTime));
        this.actions = new Actions(driver);
    }

    private By getBy(String locatorType, String locatorValue) {
        switch (locatorType.toLowerCase()) {
            case "id": return By.id(locatorValue);
            case "name": return By.name(locatorValue);
            case "xpath": return By.xpath(locatorValue);
            case "css": return By.cssSelector(locatorValue);
            case "class": return By.className(locatorValue);
            case "tag": return By.tagName(locatorValue);
            case "link_text": return By.linkText(locatorValue);
            case "partial_link_text": return By.partialLinkText(locatorValue);
            default: throw new IllegalArgumentException("Invalid locator type: " + locatorType);
        }
    }

    // --- Element Finding ---

    public WebElement findElement(String locatorType, String locatorValue) {
        return wait.until(ExpectedConditions.presenceOfElementLocated(getBy(locatorType, locatorValue)));
    }

    public WebElement findClickable(String locatorType, String locatorValue) {
        return wait.until(ExpectedConditions.elementToBeClickable(getBy(locatorType, locatorValue)));
    }

    public WebElement findVisible(String locatorType, String locatorValue) {
        return wait.until(ExpectedConditions.visibilityOfElementLocated(getBy(locatorType, locatorValue)));
    }

    // --- Actions ---

    public void click(String locatorType, String locatorValue) {
        findClickable(locatorType, locatorValue).click();
    }

    public void typeText(String locatorType, String locatorValue, String text) {
        WebElement element = findVisible(locatorType, locatorValue);
        element.clear();
        element.sendKeys(text);
    }

    public void clearField(String locatorType, String locatorValue) {
        findVisible(locatorType, locatorValue).clear();
    }

    public void selectDropdown(String locatorType, String locatorValue, String optionText) {
        WebElement element = findVisible(locatorType, locatorValue);
        new Select(element).selectByVisibleText(optionText);
    }

    public void hover(String locatorType, String locatorValue) {
        WebElement element = findVisible(locatorType, locatorValue);
        actions.moveToElement(element).perform();
    }

    public void doubleClick(String locatorType, String locatorValue) {
        WebElement element = findClickable(locatorType, locatorValue);
        actions.doubleClick(element).perform();
    }

    public void rightClick(String locatorType, String locatorValue) {
        WebElement element = findClickable(locatorType, locatorValue);
        actions.contextClick(element).perform();
    }

    // --- Navigation ---

    public void navigateTo(String url) {
        driver.get(url);
    }

    public void refresh() { driver.navigate().refresh(); }
    public void goBack() { driver.navigate().back(); }
    public void goForward() { driver.navigate().forward(); }

    // --- Getters ---

    public String getText(String locatorType, String locatorValue) {
        return findVisible(locatorType, locatorValue).getText();
    }

    public String getAttribute(String locatorType, String locatorValue, String attribute) {
        return findElement(locatorType, locatorValue).getAttribute(attribute);
    }

    public String getTitle() { return driver.getTitle(); }
    public String getCurrentUrl() { return driver.getCurrentUrl(); }

    // --- Verification ---

    public boolean isElementPresent(String locatorType, String locatorValue) {
        try {
            driver.findElement(getBy(locatorType, locatorValue));
            return true;
        } catch (NoSuchElementException e) {
            return false;
        }
    }

    public boolean isElementVisible(String locatorType, String locatorValue) {
        try {
            return driver.findElement(getBy(locatorType, locatorValue)).isDisplayed();
        } catch (Exception e) {
            return false;
        }
    }

    public void verifyText(String locatorType, String locatorValue, String expected) {
        String actual = getText(locatorType, locatorValue);
        if (!actual.contains(expected)) {
            throw new AssertionError("Text verification failed. Expected: '" + expected + "', Got: '" + actual + "'");
        }
    }

    public void verifyTitle(String expected) {
        String actual = getTitle();
        if (!actual.contains(expected)) {
            throw new AssertionError("Title verification failed. Expected: '" + expected + "', Got: '" + actual + "'");
        }
    }

    public void verifyUrlContains(String expected) {
        String actual = getCurrentUrl();
        if (!actual.contains(expected)) {
            throw new AssertionError("URL verification failed. Expected '" + expected + "' in '" + actual + "'");
        }
    }

    // --- Frames & Windows ---

    public void switchToFrame(String locatorType, String locatorValue) {
        WebElement frame = findElement(locatorType, locatorValue);
        driver.switchTo().frame(frame);
    }

    public void switchToDefaultContent() {
        driver.switchTo().defaultContent();
    }

    public void switchToWindow(int index) {
        List<String> handles = new java.util.ArrayList<>(driver.getWindowHandles());
        if (index < handles.size()) {
            driver.switchTo().window(handles.get(index));
        }
    }

    // --- Alerts ---

    public void acceptAlert() {
        wait.until(ExpectedConditions.alertIsPresent());
        driver.switchTo().alert().accept();
    }

    public void dismissAlert() {
        wait.until(ExpectedConditions.alertIsPresent());
        driver.switchTo().alert().dismiss();
    }

    // --- JavaScript ---

    public Object executeJs(String script, Object... args) {
        return ((JavascriptExecutor) driver).executeScript(script, args);
    }

    public void scrollDown(int pixels) {
        executeJs("window.scrollBy(0, " + pixels + ")");
    }

    public void scrollUp(int pixels) {
        executeJs("window.scrollBy(0, -" + pixels + ")");
    }

    public void scrollToElement(String locatorType, String locatorValue) {
        WebElement element = findElement(locatorType, locatorValue);
        executeJs("arguments[0].scrollIntoView(true);", element);
    }

    // --- Screenshot ---

    public String takeScreenshot(String name) {
        try {
            File dir = new File("../reports/screenshots");
            dir.mkdirs();
            if (name == null || name.isEmpty()) {
                name = "screenshot_" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
            }
            File screenshot = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
            File dest = new File(dir, name + ".png");
            FileUtils.copyFile(screenshot, dest);
            return dest.getAbsolutePath();
        } catch (Exception e) {
            System.err.println("Screenshot failed: " + e.getMessage());
            return null;
        }
    }

    // --- Wait ---

    public void waitSeconds(double seconds) throws InterruptedException {
        Thread.sleep((long) (seconds * 1000));
    }
}

File: python_framework/base/base_page.py
import os
import time
from datetime import datetime
from selenium.webdriver.support.ui import WebDriverWait, Select
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.keys import Keys


class BasePage:
    """Base class for all page objects. Provides common Selenium operations."""

    LOCATOR_MAP = {
        'id': By.ID,
        'name': By.NAME,
        'xpath': By.XPATH,
        'css': By.CSS_SELECTOR,
        'class': By.CLASS_NAME,
        'tag': By.TAG_NAME,
        'link_text': By.LINK_TEXT,
        'partial_link_text': By.PARTIAL_LINK_TEXT,
    }

    def __init__(self, driver, wait_time=20):
        self.driver = driver
        self.wait = WebDriverWait(driver, wait_time)
        self.actions = ActionChains(driver)

    def _get_by(self, locator_type):
        by = self.LOCATOR_MAP.get(locator_type.lower())
        if not by:
            raise ValueError(f"Invalid locator type: '{locator_type}'. "
                             f"Supported: {list(self.LOCATOR_MAP.keys())}")
        return by

    # --- Element Finding ---

    def find_element(self, locator_type, locator_value):
        by = self._get_by(locator_type)
        return self.wait.until(EC.presence_of_element_located((by, locator_value)))

    def find_elements(self, locator_type, locator_value):
        by = self._get_by(locator_type)
        return self.driver.find_elements(by, locator_value)

    def find_clickable_element(self, locator_type, locator_value):
        by = self._get_by(locator_type)
        return self.wait.until(EC.element_to_be_clickable((by, locator_value)))

    def find_visible_element(self, locator_type, locator_value):
        by = self._get_by(locator_type)
        return self.wait.until(EC.visibility_of_element_located((by, locator_value)))

    # --- Actions ---

    def click(self, locator_type, locator_value):
        element = self.find_clickable_element(locator_type, locator_value)
        element.click()

    def type_text(self, locator_type, locator_value, text):
        element = self.find_visible_element(locator_type, locator_value)
        element.clear()
        element.send_keys(text)

    def clear_field(self, locator_type, locator_value):
        element = self.find_visible_element(locator_type, locator_value)
        element.clear()

    def select_dropdown(self, locator_type, locator_value, option_text):
        element = self.find_visible_element(locator_type, locator_value)
        select = Select(element)
        select.select_by_visible_text(option_text)

    def select_dropdown_by_value(self, locator_type, locator_value, value):
        element = self.find_visible_element(locator_type, locator_value)
        select = Select(element)
        select.select_by_value(value)

    def hover(self, locator_type, locator_value):
        element = self.find_visible_element(locator_type, locator_value)
        ActionChains(self.driver).move_to_element(element).perform()

    def double_click(self, locator_type, locator_value):
        element = self.find_clickable_element(locator_type, locator_value)
        ActionChains(self.driver).double_click(element).perform()

    def right_click(self, locator_type, locator_value):
        element = self.find_clickable_element(locator_type, locator_value)
        ActionChains(self.driver).context_click(element).perform()

    def press_key(self, key_name):
        key = getattr(Keys, key_name.upper(), None)
        if key:
            ActionChains(self.driver).send_keys(key).perform()
        else:
            raise ValueError(f"Unknown key: {key_name}")

    def scroll_to_element(self, locator_type, locator_value):
        element = self.find_element(locator_type, locator_value)
        self.driver.execute_script("arguments[0].scrollIntoView(true);", element)

    def scroll_page(self, direction="down", pixels=500):
        if direction.lower() == "down":
            self.driver.execute_script(f"window.scrollBy(0, {pixels});")
        elif direction.lower() == "up":
            self.driver.execute_script(f"window.scrollBy(0, -{pixels});")

    def upload_file(self, locator_type, locator_value, file_path):
        element = self.find_element(locator_type, locator_value)
        element.send_keys(os.path.abspath(file_path))

    # --- Navigation ---

    def navigate_to(self, url):
        self.driver.get(url)

    def refresh_page(self):
        self.driver.refresh()

    def go_back(self):
        self.driver.back()

    def go_forward(self):
        self.driver.forward()

    # --- Waits ---

    def wait_for_element_visible(self, locator_type, locator_value, timeout=20):
        by = self._get_by(locator_type)
        return WebDriverWait(self.driver, timeout).until(
            EC.visibility_of_element_located((by, locator_value))
        )

    def wait_for_element_invisible(self, locator_type, locator_value, timeout=20):
        by = self._get_by(locator_type)
        return WebDriverWait(self.driver, timeout).until(
            EC.invisibility_of_element_located((by, locator_value))
        )

    def wait_seconds(self, seconds):
        time.sleep(float(seconds))

    # --- Getters ---

    def get_text(self, locator_type, locator_value):
        element = self.find_visible_element(locator_type, locator_value)
        return element.text

    def get_attribute(self, locator_type, locator_value, attribute):
        element = self.find_element(locator_type, locator_value)
        return element.get_attribute(attribute)

    def get_title(self):
        return self.driver.title

    def get_current_url(self):
        return self.driver.current_url

    # --- Verifications ---

    def is_element_present(self, locator_type, locator_value):
        try:
            self.find_element(locator_type, locator_value)
            return True
        except Exception:
            return False

    def is_element_visible(self, locator_type, locator_value):
        try:
            element = self.find_element(locator_type, locator_value)
            return element.is_displayed()
        except Exception:
            return False

    def is_element_enabled(self, locator_type, locator_value):
        try:
            element = self.find_element(locator_type, locator_value)
            return element.is_enabled()
        except Exception:
            return False

    def verify_text(self, locator_type, locator_value, expected_text):
        actual_text = self.get_text(locator_type, locator_value)
        assert expected_text in actual_text, (
            f"Text verification failed. Expected: '{expected_text}', Got: '{actual_text}'"
        )

    def verify_title(self, expected_title):
        actual_title = self.get_title()
        assert expected_title in actual_title, (
            f"Title verification failed. Expected: '{expected_title}', Got: '{actual_title}'"
        )

    def verify_url_contains(self, expected_text):
        actual_url = self.get_current_url()
        assert expected_text in actual_url, (
            f"URL verification failed. Expected '{expected_text}' in '{actual_url}'"
        )

    # --- Frames & Windows ---

    def switch_to_frame(self, locator_type=None, locator_value=None):
        if locator_type and locator_value:
            frame = self.find_element(locator_type, locator_value)
            self.driver.switch_to.frame(frame)
        else:
            self.driver.switch_to.default_content()

    def switch_to_window(self, index=1):
        handles = self.driver.window_handles
        if index < len(handles):
            self.driver.switch_to.window(handles[index])

    def close_current_window(self):
        self.driver.close()
        handles = self.driver.window_handles
        if handles:
            self.driver.switch_to.window(handles[0])

    # --- Alerts ---

    def accept_alert(self):
        self.wait.until(EC.alert_is_present())
        self.driver.switch_to.alert.accept()

    def dismiss_alert(self):
        self.wait.until(EC.alert_is_present())
        self.driver.switch_to.alert.dismiss()

    def get_alert_text(self):
        self.wait.until(EC.alert_is_present())
        return self.driver.switch_to.alert.text

    # --- JavaScript ---

    def execute_js(self, script, *args):
        return self.driver.execute_script(script, *args)

    # --- Screenshots ---

    def take_screenshot(self, name=None):
        screenshots_dir = os.path.join(os.path.dirname(__file__), '..', '..', 'reports', 'screenshots')
        os.makedirs(screenshots_dir, exist_ok=True)
        filename = name or f"screenshot_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        filepath = os.path.join(screenshots_dir, f"{filename}.png")
        self.driver.save_screenshot(filepath)
        return filepath
